File: XiePinyin/Logic/DocxExporter.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Text.RegularExpressions;
using System.IO;
using System.Reflection;
using ICSharpCode.SharpZipLib.Zip;
using System.Threading.Tasks;

namespace XiePinyin.Logic
{
    internal class DocxExporter
    {
        readonly string skDocument;
        readonly string skPara;
        readonly string skRubyWord;
        readonly string skText;
        readonly string skStyles;

        readonly Composer composer;
        readonly XieChar[] text;
        readonly string fname;

        string readStringResource(string fn)
        {
            var ass = Assembly.GetExecutingAssembly();
            using (Stream s = ass.GetManifestResourceStream("XiePinyin.Resources." + fn))
            using (StreamReader rdr = new StreamReader(s))
            {
                return rdr.ReadToEnd();
            }
        }

        public DocxExporter(Composer composer, XieChar[] text, string fname)
        {
            skDocument = readStringResource("skeleton-document.xml");
            skPara = readStringResource("skeleton-paragraph.xml");
            skRubyWord = readStringResource("skeleton-rubyword.xml");
            skText = readStringResource("skeleton-text.xml");
            skStyles = readStringResource("styles.xml");

            this.composer = composer;
            this.text = text;
            this.fname = fname;
        }

        public static void Housekeep(string folder)
        {
            // TO-DO: Have document juggler call this
            // Delete old files
            // Export file name
        }

        public class CustomStaticDataSource : IStaticDataSource
        {
            private Stream _stream;
            public Stream GetSource()
            {
                return _stream;
            }

            public void SetStream(Stream inputStream)
            {
                _stream = inputStream;
                _stream.Position = 0;
            }
        }

        void makeZip(string fn, string docXml)
        {
            var ass = Assembly.GetExecutingAssembly();
            using (Stream temaplateStream = ass.GetManifestResourceStream("XiePinyin.Resources.template.docx"))
            {
                byte[] ba = new byte[temaplateStream.Length];
                temaplateStream.Read(ba, 0, ba.Length);
                File.WriteAllBytes(fn, ba);
            }
            using (Stream zipStream = new FileStream(fn, FileMode.Open, FileAccess.ReadWrite))
            {
                ZipFile zipFile = new ZipFile(zipStream);
                zipFile.BeginUpdate();
                CustomStaticDataSource sds;
                // Update document.xml
                sds = new CustomStaticDataSource();
                sds.SetStream(new MemoryStream(Encoding.UTF8.GetBytes(docXml)));
                zipFile.Add(sds, "word/document.xml");
                // Update styles.xm
                sds = new CustomStaticDataSource();
                sds.SetStream(new MemoryStream(Encoding.UTF8.GetBytes(skStyles)));
                zipFile.Add(sds, "word/styles.xml");
                // Finish
                zipFile.CommitUpdate();
                zipFile.Close();
            }
        }

        static string esc(string str)
        {
            str = str.Replace("&", "&amp;");
            str = str.Replace("<", "&lt;");
            str = str.Replace(">", "&gt;");
            return str;
        }

        class BiWord
        {
            public string Hanzi = "";
            public string Pinyin = "";
            public bool IsEmpty { get { return Hanzi == "" && Pinyin == ""; } }
        }

        readonly Regex reWS = new Regex(@"^\s+$");

        List<BiWord> makeWords(List<XieChar> para)
        {
            var res = new List<BiWord>();
            if (para.Count == 0) return res;

            bool inAlfa = string.IsNullOrEmpty(para[0].Pinyin);
            int ix = 0;
            var word = new BiWord();
            while (ix < para.Count)
            {
                // First, eat up leading WS
                while (ix < para.Count && reWS.IsMatch(para[ix].Hanzi))
                {
                    if (inAlfa && para[ix].Pinyin != null || !inAlfa && para[ix].Pinyin == null)
                        goto WordOver;
                    word.Pinyin += " ";
                    ++ix;
                }
                if (!word.IsEmpty)
                {
                    res.Add(word);
                    word = new BiWord();
                }
                // Eat up words: non-WS followed by WS
                while (ix < para.Count)
                {
                    while (ix < para.Count && !reWS.IsMatch(para[ix].Hanzi))
                    {
                        if (inAlfa && !string.IsNullOrEmpty(para[ix].Pinyin) || !inAlfa && string.IsNullOrEmpty(para[ix].Pinyin))
                            goto WordOver;
                        if (!inAlfa)
                        {
                            if (!string.IsNullOrEmpty(para[ix].Hanzi)) word.Hanzi += para[ix].Hanzi;
                            if (!string.IsNullOrEmpty(para[ix].Pinyin)) word.Pinyin += para[ix].Pinyin;
                        }
                        else word.Pinyin += para[ix].Hanzi;
                        ++ix;
                    }
                    while (ix < para.Count && reWS.IsMatch(para[ix].Hanzi))
                    {
                        if (inAlfa && para[ix].Pinyin != null || !inAlfa && para[ix].Pinyin == null)
                            goto WordOver;
                        word.Pinyin += " ";
                        ++ix;
                    }
                    res.Add(word);
                    word = new BiWord();
                }

            WordOver:
                if (!word.IsEmpty) res.Add(word);
                word = new BiWord();
                inAlfa = !inAlfa;
            }

            // Convert Pinyin in biscriptal words to pretty accents
            foreach (var x in res)
            {
                if (x.Pinyin == "" || x.Hanzi == "") continue;
                x.Pinyin = composer.PinyinNumsToSurf(x.Pinyin);
            }

            // Done
            return res;
        }

        string makeParaXml(List<XieChar> para)
        {
            var words = makeWords(para);
            StringBuilder sb = new StringBuilder();
            foreach (var word in words)
            {
                if (word.Hanzi == "")
                    sb.Append(skText.Replace("<!-- TEXT -->", esc(word.Pinyin)));
                else
                {
                    string txt = skRubyWord.Replace("<!-- HANZI -->", esc(word.Hanzi));
                    txt = txt.Replace("<!-- PINYIN -->", esc(word.Pinyin));
                    sb.Append(txt);
                }
            }
            return sb.ToString();
        }

        static List<List<XieChar>> textToParas(XieChar[] text)
        {
            var res = new List<List<XieChar>>();
            List<XieChar> currPara = new List<XieChar>();
            foreach (var xc in text)
            {
                if (xc.Hanzi != "\n") currPara.Add(xc);
                else
                {
                    res.Add(currPara);
                    currPara = new List<XieChar>();
                }
            }
            if (currPara.Count != 0) res.Add(currPara);
            return res;
        }

        string makeDocXml(List<List<XieChar>> paras)
        {
            StringBuilder sb = new StringBuilder();
            foreach (var para in paras)
            {
                string textStr = makeParaXml(para);
                string paraStr = skPara;
                paraStr = paraStr.Replace("<!-- TEXT -->", textStr);
                sb.Append(paraStr);
            }
            return skDocument.Replace("<!-- CONTENT -->", sb.ToString());
        }

        public async Task Export()
        {
            await Task.Run(() =>
            {
                var paras = textToParas(text);
                string docXml = makeDocXml(paras);
                makeZip(fname, docXml);
            });
        }
    }
}

File: xiep/internal/docx/docxexporter.go
package docx

import (
	"archive/zip"
	"bytes"
	"embed"
	_ "embed"
	"io/ioutil"
	"os"
	"strings"
	"unicode"
	"xiep/internal/biscript"
)

//go:embed skeleton-document.xml
var skDocument string

//go:embed skeleton-paragraph.xml
var skPara string

//go:embed skeleton-rubyword.xml
var skRubyWord string

//go:embed skeleton-text.xml
var skText string

//go:embed styles.xml
var skStyles string

//go:embed template.docx
var efs embed.FS

type pinyiner interface {
	PinyinNumsToSurf(pyNums string) string
}

// Exports the received text as a DOCX file, saved as fname.
func Export(text []biscript.XieChar, fname string, pinyiner pinyiner) error {
	paras := textToParas(text)
	docXml := makeDocXml(paras, pinyiner)
	return makeZip(fname, docXml)
}

func makeZip(fname string, docXml string) error {
	// We basically copy the embedded DOCX into our output file by file
	// Except, for the document XML, we slot in our own
	fIn, err := efs.Open("template.docx")
	if err != nil {
		return err
	}
	//goland:noinspection GoUnhandledErrorResult
	defer fIn.Close()
	bytesIn, err := ioutil.ReadAll(fIn)
	if err != nil {
		return err
	}
	zipReader, err := zip.NewReader(bytes.NewReader(bytesIn), int64(len(bytesIn)))
	if err != nil {
		return err
	}
	outBuf := new(bytes.Buffer)
	zipWriter := zip.NewWriter(outBuf)

	// Read all the files from zip archive
	for _, zipFileIn := range zipReader.File {
		//fmt.Println("Reading file:", zipFileIn.Name)
		rawBytes, err := readZipFile(zipFileIn)
		if err != nil {
			return err
		}
		zipFileOut, err := zipWriter.Create(zipFileIn.Name)
		if err != nil {
			return err
		}
		// word/document.xml
		// word/styles.xml
		if zipFileIn.Name == "word/document.xml" {
			rawBytes = []byte(docXml)
		} else if zipFileIn.Name == "word/styles.xml" {
			rawBytes = []byte(skStyles)
		}
		_, err = zipFileOut.Write(rawBytes)
		if err != nil {
			return err
		}
	}
	// This is a byte buffer, not expecting an error here
	//goland:noinspection GoUnhandledErrorResult
	zipWriter.Close()
	// Now write bytes to file
	fOut, err := os.Create(fname)
	if err != nil {
		return err
	}
	_, err = fOut.Write(outBuf.Bytes())
	if err != nil {
		_ = fOut.Close()
		return err
	}
	err = fOut.Close()
	if err != nil {
		return err
	}
	// Looks like we're good.
	return nil
}

func readZipFile(zf *zip.File) ([]byte, error) {
	f, err := zf.Open()
	if err != nil {
		return nil, err
	}
	// Read-only file, no error checking needed.
	//goland:noinspection GoUnhandledErrorResult
	defer f.Close()
	return ioutil.ReadAll(f)
}

type biWord struct {
	hanzi  string
	pinyin string
}

func (w biWord) isEmpty() bool {
	return w.hanzi == "" && w.pinyin == ""
}

func isNonEmptyWS(str string) bool {
	if str == "" {
		return false
	}
	for _, r := range str {
		if !unicode.IsSpace(r) {
			return false
		}
	}
	return true
}

func makeWords(para []biscript.XieChar, pinyiner pinyiner) []biWord {
	var res []biWord
	if len(para) == 0 {
		return res
	}
	inAlpha := para[0].Pinyin == ""
	ix := 0
	var word biWord
	for ix < len(para) {
		// First, eat up leading WS
		for ix < len(para) && isNonEmptyWS(para[ix].Hanzi) {
			if inAlpha && para[ix].Pinyin != "" || !inAlpha && para[ix].Pinyin == "" {
				goto wordOver
			}
			word.pinyin += " "
			ix++
		}
		if !word.isEmpty() {
			res = append(res, word)
			word = biWord{}
		}
		// Eat up words: non-WS followed by WS
		for ix < len(para) && !isNonEmptyWS(para[ix].Hanzi) {
			if inAlpha && para[ix].Pinyin != "" || !inAlpha && para[ix].Pinyin == "" {
				goto wordOver
			}
			if !inAlpha {
				word.hanzi += para[ix].Hanzi
				word.pinyin += para[ix].Pinyin
			} else {
				word.pinyin += para[ix].Hanzi
			}
			ix++
		}
		for ix < len(para) && isNonEmptyWS(para[ix].Hanzi) {
			if inAlpha && para[ix].Pinyin != "" || !inAlpha && para[ix].Pinyin == "" {
				goto wordOver
			}
			word.pinyin += " "
			ix++
		}
		res = append(res, word)
		word = biWord{}

	wordOver:
		if !word.isEmpty() {
			res = append(res, word)
		}
		word = biWord{}
		inAlpha = !inAlpha
	}
	// Convert Pinyin in biscriptal words to pretty accents
	for ix, w := range res {
		if w.pinyin == "" || w.hanzi == "" {
			continue
		}
		res[ix].pinyin = pinyiner.PinyinNumsToSurf(w.pinyin)
	}
	// Done
	return res
}

func esc(str string) string {
	var sb strings.Builder
	sb.Grow(len(str))
	for _, c := range str {
		if c == '&' {
			sb.WriteString("&amp;")
		} else if c == '<' {
			sb.WriteString("&lt;")
		} else if c == '>' {
			sb.WriteString("&gt;")
		} else {
			sb.WriteRune(c)
		}
	}
	return sb.String()
}

func makeParaXml(para []biscript.XieChar, pinyiner pinyiner) string {
	words := makeWords(para, pinyiner)
	var sb strings.Builder
	for _, w := range words {
		if w.hanzi == "" {
			sb.WriteString(strings.ReplaceAll(skText, "<!-- TEXT -->", esc(w.pinyin)))
		} else {
			txt := strings.ReplaceAll(skRubyWord, "<!-- HANZI -->", esc(w.hanzi))
			txt = strings.ReplaceAll(txt, "<!-- PINYIN -->", esc(w.pinyin))
			sb.WriteString(txt)
		}
	}
	return sb.String()
}

func makeDocXml(paras [][]biscript.XieChar, pinyiner pinyiner) string {
	var sb strings.Builder
	for _, para := range paras {
		textStr := makeParaXml(para, pinyiner)
		paraStr := skPara
		paraStr = strings.ReplaceAll(paraStr, "<!-- TEXT -->", textStr)
		sb.WriteString(paraStr)
	}
	res := strings.ReplaceAll(skDocument, "<!-- CONTENT -->", sb.String())
	return res
}

func textToParas(text []biscript.XieChar) [][]biscript.XieChar {
	var res [][]biscript.XieChar
	var currPara []biscript.XieChar
	for _, xc := range text {
		if xc.Hanzi != "\n" {
			currPara = append(currPara, xc)
		} else {
			res = append(res, currPara)
			currPara = make([]biscript.XieChar, 0)
		}
	}
	if len(currPara) != 0 {
		res = append(res, currPara)
	}
	return res
}
